File: src/commands/recent.ts
import { ProjectIndex } from "../lib/index.ts";
import { relativeTime } from "../lib/time.ts";

export async function recent(indexPath: string, limit?: number): Promise<void> {
  const idx = await ProjectIndex.load(indexPath);
  const entries = idx.recent(limit);

  if (entries.length === 0) {
    console.error("No projects indexed. Clone a repo or run 'gx rebuild'.");
    return;
  }

  const maxName = Math.max(...entries.map(([name]) => name.length));
  for (const [name, entry] of entries) {
    const time = relativeTime(entry.lastVisited || entry.clonedAt);
    const timeStr = time ? `  ${time}` : "";
    console.log(`${name.padEnd(maxName)}  ${entry.path}${timeStr}`);
  }
}

File: crates/gx/src/commands/recent.rs
//! `gx recent [-n <count>]` — list recently visited projects with relative
//! timestamps. Ported from `src/commands/recent.ts`.

use std::path::Path;

use crate::index_store::ProjectIndex;
use crate::time::relative_time;

pub fn recent(index_path: &Path, limit: Option<usize>) {
    let idx = ProjectIndex::load(index_path);
    let entries = idx.recent(limit);
    if entries.is_empty() {
        eprintln!("No projects indexed. Clone a repo or run 'gx rebuild'.");
        return;
    }
    let max_name = entries
        .iter()
        .map(|(name, _)| name.chars().count())
        .max()
        .unwrap_or(0);
    for (name, entry) in &entries {
        // TS: relativeTime(entry.lastVisited || entry.clonedAt) — an empty
        // lastVisited is falsy and falls back to clonedAt.
        let stamp = match entry.last_visited.as_deref() {
            Some(s) if !s.is_empty() => s,
            _ => entry.cloned_at.as_str(),
        };
        let time = relative_time(stamp);
        let time_str = if time.is_empty() {
            String::new()
        } else {
            format!("  {time}")
        };
        println!("{name:<max_name$}  {}{time_str}", entry.path);
    }
}
